distribution/test/itests/test-itests-ddf/src/test/java/ddf/test/itests/catalog/TestFederation.java

```diff
@@ -1543,6 +1543,196 @@ public void testFederatedDownloadProductToCacheOnlyCacheDisabled() throws Except
+                + "?transform=resource" + "&session=" + cometDClient1.getClientId();
+
+        // Verify that product retrieval fails from the csw stub server.
+        // @formatter:off
+        when().get(restUrl).then().log().all().assertThat().statusCode(500).contentType("text/plain")
+                .body(containsString("cannot retrieve product"));
+        // @formatter:on
+
+        checkExpectations(cometDClient1, 4, 8);
+        checkExpectations(cometDClient2, 4, 8);
+
+        validateNotificationsFailure(cometDClient1, filename);
+        validateNotificationsFailure(cometDClient2, filename);
+
+        validateActivitiesFailure(cometDClient1, filename);
+        validateActivitiesFailure(cometDClient2, filename);
+
+    }
+
+    private void setupCswServer(String metacardId, String filename, String resourceData)
+            throws Exception {
+        cswServer.whenHttp()
+                .match(post("/services/csw"),
+                        withPostBodyContaining("GetRecords"),
+                        withPostBodyContaining(metacardId))
+                .then(ok(),
+                        contentType("text/xml"),
+                        bytesContent(getCswQueryResponse(metacardId).getBytes()));
+
+        cswServer.whenHttp()
+                .match(Condition.get("/services/csw"),
+                        Condition.parameter("request", "GetRecordById"),
+                        Condition.parameter("id", metacardId))
+                .then(getCswRetrievalHeaders(filename),
+                        chunkedContentWithHeaders(resourceData, Duration.ofMillis(200), 3));
+
+    }
+
+    private void checkExpectations(CometDClient cometDClient, Integer numNotifications,
+            Integer numActivities) throws Exception {
+        expect("Waiting for notifications").within(10, SECONDS)
+                .until(() -> cometDClient.getMessages(NOTIFICATIONS_CHANNEL)
+                        .size() == numNotifications);
+        expect("Waiting for activities").within(10, SECONDS)
+                .until(() -> cometDClient.getMessages(ACTIVITIES_CHANNEL)
+                        .size() == numActivities);
+
+    }
+
+    private void validateNotificationsSuccess(CometDClient cometDClient, String filename,
+            String resourceData) throws Exception {
+
+        List<String> notifications = cometDClient.getMessagesInAscOrder(NOTIFICATIONS_CHANNEL);
+        assertThat(notifications.size(), is(1));
+        CometDMessageValidator.verifyNotification(JsonPath.from(notifications.get(0)),
+                filename,
+                getResourceRetrievalCompletedMessage(resourceData.length()),
+                "complete");
+
+    }
+
+    private void validateActivitiesSuccess(CometDClient cometDClient, String filename,
+            String resourceData) throws Exception {
+        List<String> notifications = cometDClient.getMessagesInAscOrder(NOTIFICATIONS_CHANNEL);
```

Comment:
Passing in the List\<String\> already created from the desired channel (rather than passing the CometDClient) is another way you could combine these two into a single method that validates activities or notifications. Also, the activities (and possibly the notifications as well) contain the string “completed” when they’re done--I’ve been using that to check if a download was successful.